File: memex/src/__tests__/store.test.ts
import { describe, it, before, after } from "node:test";
import assert from "node:assert/strict";
import { mkdtempSync, rmSync } from "fs";
import { join } from "path";
import { tmpdir } from "os";
import { Store } from "../store.js";
import { search, context, cosineSimilarity } from "../search.js";
import { parseSource } from "../types.js";
import { bowEmbedding } from "../embedder.js";

function newTestStore(): { store: Store; dir: string } {
  const dir = mkdtempSync(join(tmpdir(), "memex-test-"));
  return { store: new Store(dir), dir };
}

describe("Store CRUD", () => {
  it("add and get", () => {
    const { store } = newTestStore();
    const id = store.add({
      content: "gRPC chosen for type safety",
      type: "decision",
      tags: ["architecture", "grpc"],
      sources: [{ project: "ai-tools", path: "src/main.ts" }],
    });
    assert.equal(id.length, 8);

    const note = store.get(id);
    assert.equal(note.content, "gRPC chosen for type safety");
    assert.equal(note.type, "decision");
    assert.equal(note.status, "open");
    assert.ok(note.created_at);
    assert.ok(note.updated_at);
  });

  it("get nonexistent throws", () => {
    const { store } = newTestStore();
    assert.throws(() => store.get("nonexistent"));
  });

  it("update", () => {
    const { store } = newTestStore();
    const id = store.add({ content: "original", type: "observation", tags: ["test"] });
    store.update(id, { content: "updated", type: "decision", tags: ["test", "updated"] });

    const note = store.get(id);
    assert.equal(note.content, "updated");
    assert.equal(note.type, "decision");
    assert.deepEqual(note.tags, ["test", "updated"]);
  });

  it("update status", () => {
    const { store } = newTestStore();
    const id = store.add({ content: "a question", type: "question" });
    store.updateStatus(id, "resolved");
    assert.equal(store.get(id).status, "resolved");
  });

  it("delete", () => {
    const { store } = newTestStore();
    const id = store.add({ content: "to delete", tags: ["deleteme"], sources: [{ project: "test", path: "foo.ts" }] });
    store.delete(id);
    assert.throws(() => store.get(id));
    assert.deepEqual(store.tagsIndex()["deleteme"], undefined);
    assert.deepEqual(store.sourcesIndex()["test:foo.ts"], undefined);
  });

  it("delete nonexistent throws", () => {
    const { store } = newTestStore();
    assert.throws(() => store.delete("nonexistent"));
  });

  it("list", () => {
    const { store } = newTestStore();
    assert.equal(store.list().length, 0);
    store.add({ content: "first note", type: "observation", tags: ["a"] });
    store.add({ content: "second note\nwith more lines", type: "decision", tags: ["b"] });
    const items = store.list();
    assert.equal(items.length, 2);
    for (const item of items) {
      assert.ok(!item.preview.includes("\n"), "preview should not contain newlines");
    }
  });
});

describe("Indexes", () => {
  it("tag index", () => {
    const { store } = newTestStore();
    const id1 = store.add({ content: "note 1", tags: ["typescript", "grpc"] });
    const id2 = store.add({ content: "note 2", tags: ["typescript", "rest"] });
    const tags = store.tagsIndex();
    assert.equal(tags["typescript"].length, 2);
    assert.deepEqual(tags["grpc"], [id1]);
    assert.deepEqual(tags["rest"], [id2]);
  });

  it("source index", () => {
    const { store } = newTestStore();
    const id1 = store.add({ content: "note about main", sources: [{ project: "myproj", path: "src/main.ts" }] });
    store.add({ content: "note about utils", sources: [{ project: "myproj", path: "src/utils.ts" }] });
    assert.deepEqual(store.sourcesIndex()["myproj:src/main.ts"], [id1]);
  });

  it("index persistence", () => {
    const dir = mkdtempSync(join(tmpdir(), "memex-test-"));
    const s1 = new Store(dir);
    const id1 = s1.add({ content: "persistent", tags: ["persist"], sources: [{ project: "proj", path: "a.ts" }] });

    const s2 = new Store(dir);
    assert.deepEqual(s2.tagsIndex()["persist"], [id1]);
    assert.deepEqual(s2.sourcesIndex()["proj:a.ts"], [id1]);
    assert.equal(s2.get(id1).content, "persistent");
  });
});

describe("Search", () => {
  it("by tag", async () => {
    const { store } = newTestStore();
    store.add({ content: "typescript note", tags: ["typescript"] });
    store.add({ content: "python note", tags: ["python"] });
    store.add({ content: "both", tags: ["typescript", "python"] });
    assert.equal((await search(store, { tag: "typescript" })).length, 2);
  });

  it("by source prefix", async () => {
    const { store } = newTestStore();
    store.add({ content: "about main", sources: [{ project: "proj", path: "src/main.ts" }] });
    store.add({ content: "about store", sources: [{ project: "proj", path: "lib/store.ts" }] });
    assert.equal((await search(store, { source: "proj:src" })).length, 1);
  });

  it("by type", async () => {
    const { store } = newTestStore();
    store.add({ content: "a decision", type: "decision" });
    store.add({ content: "an observation", type: "observation" });
    store.add({ content: "another decision", type: "decision" });
    assert.equal((await search(store, { type: "decision" })).length, 2);
  });

  it("by status", async () => {
    const { store } = newTestStore();
    const id1 = store.add({ content: "open note" });
    store.add({ content: "another open" });
    store.updateStatus(id1, "resolved");
    assert.equal((await search(store, { status: "open" })).length, 1);
  });

  it("by query (cosine similarity)", async () => {
    const { store } = newTestStore();
    const id1 = store.add({ content: "gRPC was chosen for its type safety and code generation" });
    const id2 = store.add({ content: "REST API is simpler but lacks type safety" });
    const id3 = store.add({ content: "Database migration completed successfully" });

    // Set embeddings for all notes
    store.setEmbedding(id1, bowEmbedding("gRPC was chosen for its type safety and code generation"));
    store.setEmbedding(id2, bowEmbedding("REST API is simpler but lacks type safety"));
    store.setEmbedding(id3, bowEmbedding("Database migration completed successfully"));

    const results = await search(store, { query: "type safety" });
    assert.equal(results.length, 3);
    // type safety related notes should rank higher
    assert.ok(
      results[0].content.includes("type safety") || results[1].content.includes("type safety"),
      "type safety notes should rank near top",
    );
  });

  it("combined filters", async () => {
    const { store } = newTestStore();
    store.add({ content: "typescript grpc decision", type: "decision", tags: ["typescript"] });
    store.add({ content: "typescript observation", type: "observation", tags: ["typescript"] });
    store.add({ content: "python decision", type: "decision", tags: ["python"] });
    assert.equal((await search(store, { tag: "typescript", type: "decision" })).length, 1);
  });
});

describe("Context (BFS)", () => {
  it("traverses graph", () => {
    const { store } = newTestStore();
    const id1 = store.add({ content: "main entry point", sources: [{ project: "proj", path: "src/main.ts" }] });
    const id2 = store.add({ content: "store implementation" });
    const id3 = store.add({ content: "database layer" });
    store.addRelations(id1, [{ target_id: id2, type: "relates_to" }]);
    store.addRelations(id2, [{ target_id: id3, type: "depends_on" }]);

    const results = context(store, "proj:src/main.ts", 3);
    assert.equal(results.length, 3);
  });

  it("no match returns empty", () => {
    const { store } = newTestStore();
    store.add({ content: "unrelated", sources: [{ project: "other", path: "foo.ts" }] });
    assert.equal(context(store, "nonexistent:path", 3).length, 0);
  });

  it("handles cycles", () => {
    const { store } = newTestStore();
    const id1 = store.add({ content: "note1", sources: [{ project: "proj", path: "a.ts" }] });
    const id2 = store.add({ content: "note2" });
    store.addRelations(id1, [{ target_id: id2, type: "relates_to" }]);
    store.addRelations(id2, [{ target_id: id1, type: "relates_to" }]);
    assert.equal(context(store, "proj:a.ts", 3).length, 2);
  });

  it("incoming edges are tracked", () => {
    const { store } = newTestStore();
    const id1 = store.add({ content: "note1", sources: [{ project: "proj", path: "a.ts" }] });
    const id2 = store.add({ content: "note2" });
    store.addRelations(id1, [{ target_id: id2, type: "relates_to" }]);

    const results = context(store, "proj:a.ts", 3);
    const note2 = results.find((r) => r.note.id === id2);
    assert.ok(note2?.incoming);
    assert.equal(note2!.incoming![0].target_id, id1);
  });
});

describe("Relations", () => {
  it("addRelations updates note and graph", () => {
    const { store } = newTestStore();
    const id1 = store.add({ content: "note 1" });
    const id2 = store.add({ content: "note 2" });
    store.addRelations(id1, [{ target_id: id2, type: "relates_to" }]);

    assert.equal(store.get(id1).relations.length, 1);
    assert.equal(store.graphIndex()[id1][0].target_id, id2);
  });

  it("delete cleans graph edges", () => {
    const { store } = newTestStore();
    const id1 = store.add({ content: "note 1" });
    const id2 = store.add({ content: "note 2" });
    const id3 = store.add({ content: "note 3" });
    store.addRelations(id1, [{ target_id: id2, type: "relates_to" }]);
    store.addRelations(id3, [{ target_id: id2, type: "depends_on" }]);
    store.delete(id2);

    const graph = store.graphIndex();
    assert.ok(!graph[id1] || graph[id1].length === 0);
    assert.ok(!graph[id3] || graph[id3].length === 0);
  });
});

describe("Embeddings", () => {
  it("set and get embedding", () => {
    const { store } = newTestStore();
    const id = store.add({ content: "test note" });
    assert.equal(store.getEmbedding(id), undefined);

    const vec = new Array(384).fill(0);
    vec[0] = 1.0; vec[1] = 0.5;
    store.setEmbedding(id, vec);

    const emb = store.getEmbedding(id);
    assert.ok(emb);
    assert.equal(emb![0], 1.0);
    assert.equal(emb![1], 0.5);
  });

  it("cosine similarity", () => {
    assert.ok(cosineSimilarity([1, 0, 0], [1, 0, 0]) > 0.99);
    assert.ok(Math.abs(cosineSimilarity([1, 0, 0], [0, 1, 0])) < 0.01);
    assert.ok(cosineSimilarity([1, 0, 0], [-1, 0, 0]) < -0.99);
  });
});

describe("Config", () => {
  it("defaults", () => {
    const { store } = newTestStore();
    const cfg = store.getConfig();
    assert.equal(cfg.model, "claude-haiku-4-5-20251001");
    assert.equal(cfg.embedding_enabled, true);
  });

  it("set and get", () => {
    const { store } = newTestStore();
    store.setConfig("auth_token", "test-token");
    store.setConfig("embedding_enabled", "true");
    store.setConfig("model", "claude-sonnet-4-5-20250929");

    const cfg = store.getConfig();
    assert.equal(cfg.auth_token, "test-token");
    assert.equal(cfg.embedding_enabled, true);
    assert.equal(cfg.model, "claude-sonnet-4-5-20250929");
  });

  it("unknown key throws", () => {
    const { store } = newTestStore();
    assert.throws(() => store.setConfig("unknown", "value"));
  });

  it("hook_min_turns is no longer a valid config key", () => {
    const { store } = newTestStore();
    assert.throws(() => store.setConfig("hook_min_turns", "3"), /Unknown config key/);
  });
});

describe("ID Validation", () => {
  it("rejects path traversal IDs", () => {
    const { store } = newTestStore();
    assert.throws(() => store.get("../../etc/passwd"), /invalid note ID/);
    assert.throws(() => store.get("../secret"), /invalid note ID/);
  });

  it("rejects IDs with non-hex characters", () => {
    const { store } = newTestStore();
    assert.throws(() => store.get("hello!"), /invalid note ID/);
    assert.throws(() => store.get("abc xyz"), /invalid note ID/);
    assert.throws(() => store.get("ABCDEF01"), /invalid note ID/);
  });

  it("accepts valid hex IDs", () => {
    const { store } = newTestStore();
    const id = store.add({ content: "test" });
    assert.ok(/^[a-f0-9]+$/.test(id));
    assert.doesNotThrow(() => store.get(id));
  });
});

describe("Utilities", () => {
  it("parseSource", () => {
    const src = parseSource("myproj:src/main.ts");
    assert.equal(src.project, "myproj");
    assert.equal(src.path, "src/main.ts");

    const src2 = parseSource("just/a/path.ts");
    assert.equal(src2.project, "");
    assert.equal(src2.path, "just/a/path.ts");
  });
});

File: memex/internal/memex/store_test.go
package memex

import (
	"testing"
)

func newTestStore(t *testing.T) *Store {
	t.Helper()
	dir := t.TempDir()
	s, err := NewStoreAt(dir)
	if err != nil {
		t.Fatalf("NewStoreAt failed: %v", err)
	}
	return s
}

func TestAddAndGet(t *testing.T) {
	s := newTestStore(t)

	id, err := s.Add(Note{
		Content: "gRPC chosen for type safety",
		Type:    "decision",
		Tags:    []string{"architecture", "grpc"},
		Sources: []Source{{Project: "ai-tools", Path: "cmd/main.go"}},
	})
	if err != nil {
		t.Fatalf("Add failed: %v", err)
	}
	if len(id) != 8 {
		t.Errorf("expected 8-char ID, got %q", id)
	}

	note, err := s.Get(id)
	if err != nil {
		t.Fatalf("Get failed: %v", err)
	}
	if note.Content != "gRPC chosen for type safety" {
		t.Errorf("unexpected content: %s", note.Content)
	}
	if note.Type != "decision" {
		t.Errorf("unexpected type: %s", note.Type)
	}
	if note.Status != "open" {
		t.Errorf("expected status open, got %s", note.Status)
	}
	if note.CreatedAt == "" || note.UpdatedAt == "" {
		t.Error("timestamps should be set")
	}
}

func TestGetNonexistent(t *testing.T) {
	s := newTestStore(t)

	_, err := s.Get("nonexistent")
	if err == nil {
		t.Fatal("expected error for nonexistent note")
	}
}

func TestUpdate(t *testing.T) {
	s := newTestStore(t)

	id, _ := s.Add(Note{
		Content: "original content",
		Type:    "observation",
		Tags:    []string{"test"},
	})

	err := s.Update(id, map[string]interface{}{
		"content": "updated content",
		"type":    "decision",
		"tags":    []string{"test", "updated"},
	})
	if err != nil {
		t.Fatalf("Update failed: %v", err)
	}

	note, _ := s.Get(id)
	if note.Content != "updated content" {
		t.Errorf("content not updated: %s", note.Content)
	}
	if note.Type != "decision" {
		t.Errorf("type not updated: %s", note.Type)
	}
	if len(note.Tags) != 2 || note.Tags[1] != "updated" {
		t.Errorf("tags not updated: %v", note.Tags)
	}
}

func TestUpdateStatus(t *testing.T) {
	s := newTestStore(t)

	id, _ := s.Add(Note{Content: "a question", Type: "question"})

	err := s.UpdateStatus(id, "resolved")
	if err != nil {
		t.Fatalf("UpdateStatus failed: %v", err)
	}

	note, _ := s.Get(id)
	if note.Status != "resolved" {
		t.Errorf("expected resolved, got %s", note.Status)
	}
}

func TestDelete(t *testing.T) {
	s := newTestStore(t)

	id, _ := s.Add(Note{
		Content: "to be deleted",
		Tags:    []string{"deleteme"},
		Sources: []Source{{Project: "test", Path: "foo.go"}},
	})

	err := s.Delete(id)
	if err != nil {
		t.Fatalf("Delete failed: %v", err)
	}

	_, err = s.Get(id)
	if err == nil {
		t.Fatal("expected error after delete")
	}

	// Verify indexes are cleaned up
	tags := s.TagsIndex()
	if ids, ok := tags["deleteme"]; ok && len(ids) > 0 {
		t.Error("tag index not cleaned up")
	}

	sources := s.SourcesIndex()
	if ids, ok := sources["test:foo.go"]; ok && len(ids) > 0 {
		t.Error("source index not cleaned up")
	}
}

func TestDeleteNonexistent(t *testing.T) {
	s := newTestStore(t)

	err := s.Delete("nonexistent")
	if err == nil {
		t.Fatal("expected error for nonexistent delete")
	}
}

func TestList(t *testing.T) {
	s := newTestStore(t)

	// Empty list
	list, err := s.List()
	if err != nil {
		t.Fatalf("List failed: %v", err)
	}
	if len(list) != 0 {
		t.Errorf("expected empty list, got %d items", len(list))
	}

	// Add notes
	s.Add(Note{Content: "first note", Type: "observation", Tags: []string{"a"}})
	s.Add(Note{Content: "second note\nwith more lines", Type: "decision", Tags: []string{"b"}})

	list, err = s.List()
	if err != nil {
		t.Fatalf("List failed: %v", err)
	}
	if len(list) != 2 {
		t.Errorf("expected 2 items, got %d", len(list))
	}

	// Verify preview truncation
	for _, item := range list {
		if item.Preview == "second note\nwith more lines" {
			t.Error("preview should be truncated to first line")
		}
	}
}

func TestTagIndex(t *testing.T) {
	s := newTestStore(t)

	id1, _ := s.Add(Note{Content: "note 1", Tags: []string{"go", "grpc"}})
	id2, _ := s.Add(Note{Content: "note 2", Tags: []string{"go", "rest"}})

	tags := s.TagsIndex()

	goIDs := tags["go"]
	if len(goIDs) != 2 {
		t.Errorf("expected 2 notes with 'go' tag, got %d", len(goIDs))
	}

	grpcIDs := tags["grpc"]
	if len(grpcIDs) != 1 || grpcIDs[0] != id1 {
		t.Errorf("unexpected grpc tag index: %v", grpcIDs)
	}

	restIDs := tags["rest"]
	if len(restIDs) != 1 || restIDs[0] != id2 {
		t.Errorf("unexpected rest tag index: %v", restIDs)
	}
}

func TestSourceIndex(t *testing.T) {
	s := newTestStore(t)

	id1, _ := s.Add(Note{
		Content: "note about main",
		Sources: []Source{{Project: "myproj", Path: "cmd/main.go"}},
	})
	s.Add(Note{
		Content: "note about utils",
		Sources: []Source{{Project: "myproj", Path: "internal/utils.go"}},
	})

	sources := s.SourcesIndex()

	mainIDs := sources["myproj:cmd/main.go"]
	if len(mainIDs) != 1 || mainIDs[0] != id1 {
		t.Errorf("unexpected source index for main: %v", mainIDs)
	}
}

func TestSearchByTag(t *testing.T) {
	s := newTestStore(t)

	s.Add(Note{Content: "go note", Tags: []string{"go"}})
	s.Add(Note{Content: "python note", Tags: []string{"python"}})
	s.Add(Note{Content: "both", Tags: []string{"go", "python"}})

	results, err := s.Search(SearchParams{Tag: "go"})
	if err != nil {
		t.Fatalf("Search failed: %v", err)
	}
	if len(results) != 2 {
		t.Errorf("expected 2 results for 'go' tag, got %d", len(results))
	}
}

func TestSearchBySource(t *testing.T) {
	s := newTestStore(t)

	s.Add(Note{
		Content: "about cmd",
		Sources: []Source{{Project: "proj", Path: "cmd/main.go"}},
	})
	s.Add(Note{
		Content: "about internal",
		Sources: []Source{{Project: "proj", Path: "internal/store.go"}},
	})

	results, err := s.Search(SearchParams{Source: "proj:cmd"})
	if err != nil {
		t.Fatalf("Search failed: %v", err)
	}
	if len(results) != 1 {
		t.Errorf("expected 1 result for source prefix, got %d", len(results))
	}
}

func TestSearchByType(t *testing.T) {
	s := newTestStore(t)

	s.Add(Note{Content: "a decision", Type: "decision"})
	s.Add(Note{Content: "an observation", Type: "observation"})
	s.Add(Note{Content: "another decision", Type: "decision"})

	results, err := s.Search(SearchParams{Type: "decision"})
	if err != nil {
		t.Fatalf("Search failed: %v", err)
	}
	if len(results) != 2 {
		t.Errorf("expected 2 decisions, got %d", len(results))
	}
}

func TestSearchByStatus(t *testing.T) {
	s := newTestStore(t)

	id1, _ := s.Add(Note{Content: "open note"})
	s.Add(Note{Content: "another open note"})
	s.UpdateStatus(id1, "resolved")

	results, err := s.Search(SearchParams{Status: "open"})
	if err != nil {
		t.Fatalf("Search failed: %v", err)
	}
	if len(results) != 1 {
		t.Errorf("expected 1 open note, got %d", len(results))
	}
}

func TestSearchByQuery(t *testing.T) {
	s := newTestStore(t)

	s.Add(Note{Content: "gRPC was chosen for its type safety and code generation"})
	s.Add(Note{Content: "REST API is simpler but lacks type safety"})
	s.Add(Note{Content: "Database migration completed successfully"})

	results, err := s.Search(SearchParams{Query: "type safety"})
	if err != nil {
		t.Fatalf("Search failed: %v", err)
	}
	if len(results) < 2 {
		t.Errorf("expected at least 2 results for 'type safety', got %d", len(results))
	}
	// BM25 should rank the more relevant results higher
	if len(results) >= 2 {
		// Both gRPC and REST notes mention "type safety"
		foundGRPC := false
		foundREST := false
		for _, r := range results {
			if r.Content == "gRPC was chosen for its type safety and code generation" {
				foundGRPC = true
			}
			if r.Content == "REST API is simpler but lacks type safety" {
				foundREST = true
			}
		}
		if !foundGRPC || !foundREST {
			t.Error("expected both type safety notes in results")
		}
	}
}

func TestSearchCombinedFilters(t *testing.T) {
	s := newTestStore(t)

	s.Add(Note{Content: "go grpc decision", Type: "decision", Tags: []string{"go"}})
	s.Add(Note{Content: "go observation", Type: "observation", Tags: []string{"go"}})
	s.Add(Note{Content: "python decision", Type: "decision", Tags: []string{"python"}})

	results, err := s.Search(SearchParams{Tag: "go", Type: "decision"})
	if err != nil {
		t.Fatalf("Search failed: %v", err)
	}
	if len(results) != 1 {
		t.Errorf("expected 1 result for go+decision, got %d", len(results))
	}
}

func TestContextBFS(t *testing.T) {
	s := newTestStore(t)

	// Create a small graph:
	// note1 (source: proj:cmd/main.go) --relates_to--> note2 --depends_on--> note3
	id1, _ := s.Add(Note{
		Content: "main entry point",
		Sources: []Source{{Project: "proj", Path: "cmd/main.go"}},
	})
	id2, _ := s.Add(Note{Content: "store implementation"})
	id3, _ := s.Add(Note{Content: "database layer"})

	s.AddRelations(id1, []Relation{{TargetID: id2, Type: "relates_to"}})
	s.AddRelations(id2, []Relation{{TargetID: id3, Type: "depends_on"}})

	results, err := s.Context("proj:cmd/main.go", 3)
	if err != nil {
		t.Fatalf("Context failed: %v", err)
	}

	// Should find all 3 notes via BFS
	if len(results) != 3 {
		t.Errorf("expected 3 notes in context, got %d", len(results))
		for _, r := range results {
			t.Logf("  - %s: %s", r.Note.ID, r.Note.Content)
		}
	}

	// Verify note1 has no incoming edges from our perspective
	// note2 should have incoming from note1
	// note3 should have incoming from note2
	for _, r := range results {
		if r.Note.ID == id2 {
			if len(r.Incoming) != 1 || r.Incoming[0].TargetID != id1 {
				t.Errorf("note2 should have incoming edge from note1")
			}
		}
		if r.Note.ID == id3 {
			if len(r.Incoming) != 1 || r.Incoming[0].TargetID != id2 {
				t.Errorf("note3 should have incoming edge from note2")
			}
		}
	}
}

func TestContextNoMatch(t *testing.T) {
	s := newTestStore(t)

	s.Add(Note{
		Content: "unrelated",
		Sources: []Source{{Project: "other", Path: "foo.go"}},
	})

	results, err := s.Context("nonexistent:path", 3)
	if err != nil {
		t.Fatalf("Context failed: %v", err)
	}
	if len(results) != 0 {
		t.Errorf("expected 0 results for nonexistent source, got %d", len(results))
	}
}

func TestContextCycleHandling(t *testing.T) {
	s := newTestStore(t)

	// Create a cycle: note1 → note2 → note1
	id1, _ := s.Add(Note{
		Content: "note1",
		Sources: []Source{{Project: "proj", Path: "a.go"}},
	})
	id2, _ := s.Add(Note{Content: "note2"})

	s.AddRelations(id1, []Relation{{TargetID: id2, Type: "relates_to"}})
	s.AddRelations(id2, []Relation{{TargetID: id1, Type: "relates_to"}})

	results, err := s.Context("proj:a.go", 3)
	if err != nil {
		t.Fatalf("Context failed: %v", err)
	}
	if len(results) != 2 {
		t.Errorf("expected 2 notes (cycle should not cause duplicates), got %d", len(results))
	}
}

func TestAddRelations(t *testing.T) {
	s := newTestStore(t)

	id1, _ := s.Add(Note{Content: "note 1"})
	id2, _ := s.Add(Note{Content: "note 2"})

	err := s.AddRelations(id1, []Relation{
		{TargetID: id2, Type: "relates_to"},
	})
	if err != nil {
		t.Fatalf("AddRelations failed: %v", err)
	}

	note, _ := s.Get(id1)
	if len(note.Relations) != 1 {
		t.Errorf("expected 1 relation, got %d", len(note.Relations))
	}
	if note.Relations[0].TargetID != id2 {
		t.Errorf("unexpected relation target: %s", note.Relations[0].TargetID)
	}

	// Verify graph index
	graph := s.GraphIdx()
	edges := graph[id1]
	if len(edges) != 1 || edges[0].TargetID != id2 {
		t.Errorf("graph index not updated correctly: %v", edges)
	}
}

func TestEmbedding(t *testing.T) {
	s := newTestStore(t)

	id, _ := s.Add(Note{Content: "test note"})

	// No embedding initially
	emb := s.GetEmbedding(id)
	if emb != nil {
		t.Error("expected nil embedding initially")
	}

	// Set embedding
	vec := make([]float32, 384)
	vec[0] = 1.0
	vec[1] = 0.5
	err := s.SetEmbedding(id, vec)
	if err != nil {
		t.Fatalf("SetEmbedding failed: %v", err)
	}

	// Get embedding
	emb = s.GetEmbedding(id)
	if emb == nil {
		t.Fatal("expected non-nil embedding")
	}
	if emb[0] != 1.0 || emb[1] != 0.5 {
		t.Errorf("unexpected embedding values: %v", emb[:5])
	}
}

func TestCosineSimilarity(t *testing.T) {
	// Same vector → similarity = 1.0
	a := []float32{1, 0, 0}
	sim := CosineSimilarity(a, a)
	if sim < 0.99 {
		t.Errorf("expected ~1.0 for identical vectors, got %f", sim)
	}

	// Orthogonal vectors → similarity = 0.0
	b := []float32{0, 1, 0}
	sim = CosineSimilarity(a, b)
	if sim > 0.01 {
		t.Errorf("expected ~0.0 for orthogonal vectors, got %f", sim)
	}

	// Opposite vectors → similarity = -1.0
	c := []float32{-1, 0, 0}
	sim = CosineSimilarity(a, c)
	if sim > -0.99 {
		t.Errorf("expected ~-1.0 for opposite vectors, got %f", sim)
	}
}

func TestConfig(t *testing.T) {
	s := newTestStore(t)

	// Default config
	cfg := s.GetConfig()
	if cfg.Model != "claude-haiku-4-5-20251001" {
		t.Errorf("unexpected default model: %s", cfg.Model)
	}
	if cfg.EmbeddingEnabled {
		t.Error("embedding should be disabled by default")
	}

	// Set config values
	err := s.SetConfig("api_key", "test-key")
	if err != nil {
		t.Fatalf("SetConfig failed: %v", err)
	}
	err = s.SetConfig("embedding_enabled", "true")
	if err != nil {
		t.Fatalf("SetConfig failed: %v", err)
	}
	err = s.SetConfig("model", "claude-sonnet-4-5-20250929")
	if err != nil {
		t.Fatalf("SetConfig failed: %v", err)
	}

	cfg = s.GetConfig()
	if cfg.APIKey != "test-key" {
		t.Errorf("api_key not set: %s", cfg.APIKey)
	}
	if !cfg.EmbeddingEnabled {
		t.Error("embedding_enabled not set")
	}
	if cfg.Model != "claude-sonnet-4-5-20250929" {
		t.Errorf("model not set: %s", cfg.Model)
	}

	// Unknown key
	err = s.SetConfig("unknown_key", "value")
	if err == nil {
		t.Error("expected error for unknown config key")
	}
}

func TestDeleteCleansGraphEdges(t *testing.T) {
	s := newTestStore(t)

	id1, _ := s.Add(Note{Content: "note 1"})
	id2, _ := s.Add(Note{Content: "note 2"})
	id3, _ := s.Add(Note{Content: "note 3"})

	// Create edges: id1 → id2, id3 → id2
	s.AddRelations(id1, []Relation{{TargetID: id2, Type: "relates_to"}})
	s.AddRelations(id3, []Relation{{TargetID: id2, Type: "depends_on"}})

	// Delete id2 — should clean up all edges to/from it
	err := s.Delete(id2)
	if err != nil {
		t.Fatalf("Delete failed: %v", err)
	}

	graph := s.GraphIdx()
	// id1's edge to id2 should be gone
	if edges, ok := graph[id1]; ok && len(edges) > 0 {
		t.Errorf("edges from id1 should be cleaned up: %v", edges)
	}
	// id3's edge to id2 should be gone
	if edges, ok := graph[id3]; ok && len(edges) > 0 {
		t.Errorf("edges from id3 should be cleaned up: %v", edges)
	}
}

func TestIndexPersistence(t *testing.T) {
	dir := t.TempDir()

	// Create store, add notes
	s1, _ := NewStoreAt(dir)
	id1, _ := s1.Add(Note{
		Content: "persistent note",
		Tags:    []string{"persist"},
		Sources: []Source{{Project: "proj", Path: "a.go"}},
	})

	// Create new store at same dir — should load indexes from disk
	s2, _ := NewStoreAt(dir)

	tags := s2.TagsIndex()
	if ids, ok := tags["persist"]; !ok || len(ids) != 1 || ids[0] != id1 {
		t.Errorf("tags index not persisted: %v", tags)
	}

	sources := s2.SourcesIndex()
	if ids, ok := sources["proj:a.go"]; !ok || len(ids) != 1 || ids[0] != id1 {
		t.Errorf("sources index not persisted: %v", sources)
	}

	// Note should be readable
	note, err := s2.Get(id1)
	if err != nil {
		t.Fatalf("Get from reloaded store failed: %v", err)
	}
	if note.Content != "persistent note" {
		t.Errorf("unexpected content from reloaded store: %s", note.Content)
	}
}

func TestKeywordOverlapScore(t *testing.T) {
	score := KeywordOverlapScore("gRPC type safety", "REST type safety")
	if score < 0.3 {
		t.Errorf("expected significant overlap, got %f", score)
	}

	score = KeywordOverlapScore("gRPC", "database migration")
	if score > 0.1 {
		t.Errorf("expected low overlap, got %f", score)
	}
}

func TestDetectProject(t *testing.T) {
	src := DetectProject("myproj:cmd/main.go")
	if src.Project != "myproj" || src.Path != "cmd/main.go" {
		t.Errorf("unexpected: %+v", src)
	}

	src = DetectProject("just/a/path.go")
	if src.Project != "" || src.Path != "just/a/path.go" {
		t.Errorf("unexpected: %+v", src)
	}
}

func TestBowEmbedding(t *testing.T) {
	emb := bowEmbedding("hello world test")
	if len(emb) != 384 {
		t.Errorf("expected 384-dim embedding, got %d", len(emb))
	}

	// Should be normalized (L2 norm ≈ 1.0)
	var norm float64
	for _, v := range emb {
		norm += float64(v) * float64(v)
	}
	if norm < 0.99 || norm > 1.01 {
		t.Errorf("expected normalized vector (norm ≈ 1.0), got %f", norm)
	}

	// Same input should produce same embedding
	emb2 := bowEmbedding("hello world test")
	sim := CosineSimilarity(emb, emb2)
	if sim < 0.99 {
		t.Errorf("same input should produce identical embedding, similarity = %f", sim)
	}
}
